src/jit/gentree.cpp

```diff
@@ -3132,77 +3132,134 @@ bool GenTree::gtIsValid64RsltMul()
 
 #endif // DEBUG
 
-/*****************************************************************************
- *
- *  Figure out the evaluation order for a list of values.
- */
+ //------------------------------------------------------------------------------
+ // gtSetListOrder : Figure out the evaluation order for a list of values.
+ //
+ //
+ // Arguments:
+ //    list  - List to figure out the evaluation order for
+ //    isListCallArgs - True iff the list is a list of call arguments
+ //    callArgsInRegs -  True iff the list is a list of call arguments and they are passed in registers
+ //
+ // Return Value:
+ //    True if the operation can be a root of a bitwise rotation tree; false otherwise.
 
-unsigned Compiler::gtSetListOrder(GenTree* list, bool regs)
+unsigned            Compiler::gtSetListOrder(GenTree *list, bool isListCallArgs, bool callArgsInRegs)
 {
-    assert(list && list->IsList());
+    assert((list != nullptr) && list->IsList());
+    assert(!callArgsInRegs || isListCallArgs);
 
-    unsigned level  = 0;
-    unsigned ftreg  = 0;
-    unsigned costSz = 0;
-    unsigned costEx = 0;
+    ArrayStack<GenTree *> listNodes(this);
 
+    do
+    {
+        listNodes.Push(list);
+        list = list->gtOp.gtOp2;
+    } while ((list != nullptr) && (list->IsList()));
+
+    unsigned  nxtlvl = (list == nullptr) ? 0 : gtSetEvalOrder(list);
+    while (listNodes.Height() > 0)
+    {
 #if FEATURE_STACK_FP_X87
-    /* Save the current FP stack level since an argument list
-     * will implicitly pop the FP stack when pushing the argument */
-    unsigned FPlvlSave = codeGen->genGetFPstkLevel();
+        /* Save the current FP stack level since an argument list
+        * will implicitly pop the FP stack when pushing the argument */
+        unsigned        FPlvlSave = codeGen->genGetFPstkLevel();
 #endif // FEATURE_STACK_FP_X87
 
-    GenTreePtr next = list->gtOp.gtOp2;
+        list = listNodes.Pop();
+        assert(list && list->IsList());
+        GenTreePtr      next = list->gtOp.gtOp2;
 
-    if (next)
-    {
-        unsigned nxtlvl = gtSetListOrder(next, regs);
+        unsigned        level = 0;
+        unsigned        ftreg = 0;
 
-        ftreg |= next->gtRsvdRegs;
+        // TODO: Do we have to compute costs differently for argument lists and
```

Comment:
These cost are just a best estimate of  code size and execution time.

It can impact things such a places in the JIT where we would like to duplicate a piece of code.
The JIT will do so if the cost is low and omit the duplication if the cost is too high.

Other than that the costs can be off by a lot without having much impact on what code the JIT generates.